### PR TITLE
C++: improve performance of overlap computation

### PR DESCRIPTION
**Pull request by rdmarsh2**
This is still quadratic in the number of MemoryLocations for a vvar, but
only for a single pipeline step, which is not materialized. It seems to be
fast enough in practice for the IR.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll [hunk 1]
```diff
@@ -234,6 +229,50 @@ Overlap getOverlap(MemoryLocation def, MemoryLocation use) {
   )
 }
 
+predicate isRelevantOffset(VirtualVariable vv, IntValue offset) {
```
Comment (dave-bartolomeo) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: These predicates should be private.

Comment (dave-bartolomeo) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Does this predicate have to hold even for unknown offsets?

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll [hunk 2]
```diff
@@ -234,6 +229,50 @@ Overlap getOverlap(MemoryLocation def, MemoryLocation use) {
...
+  vml.getVirtualVariable() = vv and
+  (
+    vml.getStartBitOffset() = Ints::unknown() or
+    vml.getEndBitOffset() = Ints::unknown()
+  )
+}
+
+
+
+Overlap getVariableMemoryLocationOverlap(VariableMemoryLocation def, VariableMemoryLocation use) {
```
Comment (dave-bartolomeo) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Please add a comment describing why you've implemented it this way for performance reasons.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll (hunk 1 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: It doesn't

**Pull request merged**